traits/observers/_observer_change_notifier.py

```diff
@@ -0,0 +1,167 @@
+        if isinstance(handler, types.MethodType):
+            self.handler = weakref.WeakMethod(handler)
+        else:
+            self.handler = partial(_return, value=handler)
+        self.dispatcher = dispatcher
+
+    def add_to(self, observable):
+        """ Add this notifier to the observable.
+
+        Parameters
+        ----------
+        observable : IObservable
+        """
+        notifiers = observable._notifiers(True)
+        for notifier in notifiers:
+            if self.equals(notifier):
+                break
+        else:
+            notifiers.append(self)
+
+    def remove_from(self, observable):
+        """ Remove a notifier equivalent to this one from the observable.
+
+        Parameters
+        ----------
+        observable : IObservable
+
+        Raises
+        ------
+        ValueError
+            If the notifier cannot be found.
+        """
+        notifiers = observable._notifiers(True)
+        for notifier in notifiers[:]:
+            if self.equals(notifier):
+                notifiers.remove(notifier)
+                break
+        else:
+            raise ValueError("Notifier not found.")
+
+    def __call__(self, *args, **kwargs):
+        """ Called by the observable this notifier is attached to.
+
+        This exercises the observer_handler for maintaining observers and
+        notifiers on changed objects.
+
+        Note that any unexpected exceptions will be raised, as the
+        ``observer_handler`` is not provided by users of traits, but is
+        a callable maintained in traits.
+        """
+        target = self.target()
+        if target is None:
+            return
+
+        handler = self.handler()
+        if handler is None:
+            return
+
+        event = self.event_factory(*args, **kwargs)
+
+        self.observer_handler(
+            event=event,
+            path=self.path,
+            target=target,
+            handler=handler,
+            dispatcher=self.dispatcher,
+        )
+
+    def equals(self, other):
+        """ Return true if the other value is a notifier equivalent to this one.
+
+        Parameters
+        ----------
+        other : any
+
+        Returns
+        -------
+        boolean
+        """
+        return (
+            type(self) is type(other)
+            # observer_handler contains the logic for maintaining notifiers
+            # in the downstream path.
+            and self.observer_handler is other.observer_handler
+            # path is an input for observer_handler.
+            # Unequal paths should not interfere each other.
+            and self.path == other.path
+            # user handler is an input for observer_handler.
+            # different user handlers should not interfere each other.
+            and self.handler() is other.handler()
+            # target is an input for observer_handler.
+            # it goes together with the user's handler
+            and self.target() is other.target()
```

Comment:
Same question as in #1000, should we differentiate notifiers with different dispatch?  The existing on_trait_change does not, but here is an opportunity to change that.

Reply:
It is difficult to see why these checks are needed out of the larger, integrated context.
I added a few tests on the PoC branch to convince myself that these checks are necessary: https://github.com/enthought/traits/compare/f34fa3920d41ec6cbe9d513cefcc35fd5c7edc65...986fcd25fc68f867d4e4421eba90dc7619231033

Reply:
In my test branch where I have cleaned up things, the last test shown in the diff does not fail when `self.target() is other.target()` is removed any more. 
But this one does (using the `observer-framework-test` branch). The last assertion would have failed with `1 != 0` if targets are not compared.

<details>
<summary> A test that justifies checking target identity </summary>

```

class Potato(HasTraits):

    name = Str()


class PotatoBag(HasTraits):

    potatos = List(Instance(Potato))


class Crate(HasTraits):

    potato_bags = List(PotatoBag)


class TestHasTraitsObserverDifferentiateParent(unittest.TestCase):

    def test_shared_instance_same_graph_different_target(self):

        crate1 = Crate()
        crate2 = Crate()

        # given
        handler = mock.Mock()
        crate1.observe(handler, "potato_bags:items:potatos:items:name")
        crate2.observe(handler, "potato_bags:items:potatos:items:name")

        new_potato = Potato()
        new_potato_bag = PotatoBag(potatos=[new_potato])
        crate1.potato_bags = [new_potato_bag]
        crate2.potato_bags = [new_potato_bag]
        new_potato.name = "King Edward I"
        self.assertEqual(handler.call_count, 2)
        handler.reset_mock()

        # when
        # remove the second observer
        crate2.observe(
            handler, "potato_bags:items:potatos:items:name", remove=True)
        new_potato.name = "King Edward II"

        # then
        self.assertEqual(handler.call_count, 1)
        handler.reset_mock()

        # then
        # This check the observer is maintained.
        maris_piper = Potato()
        crate2.potato_bags[0].potatos.append(maris_piper)
        crate1.potato_bags = []
        self.assertEqual(handler.call_count, 0)  # sanity check

        maris_piper.name = "Maris Piper"
        self.assertEqual(handler.call_count, 0)
```

</details>

Reply:
> Same question as in #1000, should we differentiate notifiers with different dispatch? The existing on_trait_change does not, but here is an opportunity to change that.

I am going to add the comparison on the `dispatcher` here... see discussion in #1000.